File: Find First and Last Position of Element in Sorted Array/Find_First_and_Last_Position_of_Element_in_Sorted_Array.py
# 使用两次二分查找分别对起始点和终止点进行搜索
# 和search in a 2D matrix 很相似，考察的都是二分查找的变体
# 40ms 95.13%
class Solution:
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        if not nums:
            return [-1, -1]
        
        # 查找第一个等于目标值的索引位置
        left, right = 0, len(nums) - 1
        while right - left > 1:
            mid = (left + right) // 2
            if nums[mid] >= target:
                right = mid
            else:
                left = mid + 1
        if nums[left] == target:
            start_index = left
        elif nums[right] == target:
            start_index = right
        else:
            return [-1, -1]
        # 查找最后一个等于目标值的索引位置
        left, right = start_index, len(nums) - 1
        while right - left > 1:
            mid = (left + right) // 2
            if nums[mid] <= target:
                left = mid
            else:
                right = mid - 1
        if nums[right] == target:
            return [start_index, right]
        elif nums[left] == target:
            return [start_index, left]

File: Find First and Last Position of Element in Sorted Array/Find_First_and_Last _Position_of_Element _in_Sorted_Array.cpp
// 类似的问题是在一个排好序的数组中，统计给定的一个数字的出现次数
// 大概思路都是二分查找

// 这道题很关键，因为二分查找到最后只有两个数字的时候是有可能造成无限循环的

// Runtime: 8 ms, faster than 99.66% of C++ online submissions for Find First and Last Position of Element in Sorted Array.
// Memory Usage: 10.3 MB, less than 100.00% of C++ online submissions for Find First and Last Position of Element in Sorted Array.

class Solution 
{
public:
    vector<int> searchRange(vector<int>& nums, int target) 
    {
        vector<int> res(2, -1);
        if (nums.empty())
            return res;

        // 第一次二分查找，这里为了减少每一次循环时的判断数量，将三种情况合并为两种
        int left = 0, right = nums.size() - 1;
        while(left < right)
        {
            int mid = (left + right) / 2;
            if (nums[mid] < target)
                left = mid + 1;
            else
                right = mid;
        }

        if (nums[left] == target)
            res[0] = left;
        else
            return res;
        
        // 第二次二分查找        
        right = nums.size() - 1;
        while(left < right)
        {
            // 无限循环的解决方法！！！
            int mid = (left + right) / 2 + 1;
            if (nums[mid] <= target)
                left = mid;
            else
                right = mid - 1;
        }
        res[1] = right;
        return res;
    }
};